### PR TITLE
ForbiddenAnnotation rule #2719

### PR DESCRIPTION
**Pull request by ov7a**
As discussed in #2719, this rule could be sometimes useful for cases when an annotation does not require import: `java.lang.SuppressWarnings`, `kotlin.jvm.Transient`, etc. These cases can't be caught with the `ForbiddenImport` rule.
Also, this rule can be useful for pushing people towards the Kotlin implementations (see [this issue](https://github.com/jhipster/jhipster-kotlin/issues/143) for an example).

More specifically, forbidding `kotlin.jvm.Transient` can be useful in applications with `kotlinx.serialization`: `kotlin.jvm.Transient` does not affect `@Serializable` classes. `ForbiddenAnnotation` can be a quick-fix for that.

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    fun `should report nothing when annotations are blank`() {
+        val code = """
+        @SuppressWarnings("unused") 
+        fun main() {}
+        """.trimIndent()
+        val findings = ForbiddenAnnotation(
+            TestConfig(mapOf(ANNOTATIONS to listOf("  ")))
+        ).compileAndLintWithContext(env, code)
+        assertThat(findings).isEmpty()
+    }
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:45]: You can remove this test. I imagine this is copied from `ForbiddenImport` and we had it there for "historical reasons". But it is not needed here.

**Review by BraisGabin — APPROVED** [review]
Good contribution! I really like this one!

I'm blocking this PR because we are really near to cut a new version and it's a bit late to add a new rule to it after 3 RC but it should be inside the next version :)

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt (hunk 1 above)
Comment (ov7a) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:45]: Yep, exactly! Thank you!

File: detekt-core/src/main/resources/default-detekt-config.yml [hunk 2]
```diff
@@ -528,6 +528,11 @@ style:
   ExpressionBodySyntax:
     active: false
     includeLineWrapping: false
+  ForbiddenAnnotation:
+    active: false
+    annotations:
+      - reason: 'it is a java annotation. Use `Suppress` instead.'
+        value: 'java.lang.SuppressWarnings'
```
Comment (TWiStErRob) [inline detekt-core/src/main/resources/default-detekt-config.yml:535]: Maybe add similar defaults for Java annotation annotations?

i.e. java.lang.annotation.* (Retention/Target for sure, maybe Documented/Inherited/Repeatable) most of these have their corresponding `kotlin.annotation.*` equivalent.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ */
+@RequiresTypeResolution
+class ForbiddenAnnotation(config: Config = Config.empty) : Rule(config) {
+
+    override val issue = Issue(
+        javaClass.simpleName,
+        Severity.Style,
+        "Mark forbidden annotations. A forbidden annotation could be an invocation of an unwanted " +
+            "language annotation which does not require explicit import and hence you might want to mark it " +
+            "as forbidden in order to get warned about the usage.",
```
Comment (TWiStErRob) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:38]: Careful with lengthy descriptions like this, see https://github.com/detekt/detekt/issues/4750 why it could be a problem. Shorten to a concise description and move other info, like how to action it to the KDoc on the class.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    )
+
+    @Configuration(
+        "List of fully qualified annotation classes which are forbidden. " +
+            "For example, `kotlin.jvm.Transient`."
+    )
+    private val annotations: Map<String, Forbidden> by config(
+        valuesWithReason(
+            "java.lang.SuppressWarnings" to "it is a java annotation. Use `Suppress` instead.",
+        )
```
Comment (TWiStErRob) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:49]: is this going to be the default reason for any annotation?

File: detekt-core/src/main/resources/default-detekt-config.yml (hunk 2 above)
Comment (TWiStErRob) [reply detekt-core/src/main/resources/default-detekt-config.yml:535]: + Deprecated

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+import io.gitlab.arturbosch.detekt.test.TestConfig
+import io.gitlab.arturbosch.detekt.test.assertThat
+import io.gitlab.arturbosch.detekt.test.compileAndLintWithContext
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.junit.jupiter.api.Test
+
+private const val ANNOTATIONS = "annotations"
+
+@KotlinCoreEnvironmentTest
+class ForbiddenAnnotationSpec(val env: KotlinCoreEnvironment) {
```
Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:14]: add test for nested annotations, e.g. forbid `ReplaceWith` only, and test that it's caught when used inside kotlin's Deprecated.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            annotations[it.asString()]
+        }
+
+        if (forbidden != null) {
+            val message = if (forbidden.reason != null) {
+                "The annotation `${forbidden.name}` has been forbidden: ${forbidden.reason}"
+            } else {
+                "The annotation `${forbidden.name}` has been forbidden in the detekt config."
+            }
+            report(CodeSmell(issue, Entity.from(annotation), message))
```
Comment (TWiStErRob) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:70]: please make sure you flag only the annotation, not the parameters, i.e.
```
@Deprecated(message = "abc")
^^^^^^^^^^^
```

**Comment by TWiStErRob** [comment]
@BraisGabin we can unblock this, right? Unless we need to keep `main` clean for a few days/weeks for 1.22.1?

**Comment by BraisGabin** [comment]
No need to block it any longer.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt (hunk 4 above)
Comment (ov7a) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:49]: I think not. For example, `@Inherited` isn't supported by Kotlin at all https://youtrack.jetbrains.com/issue/KT-22265

Comment (TWiStErRob) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt]: aaaah, this is the default param for a config :) so the default.yml will be generated from here I guess.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
```
Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:14]: add a test for aliased annotations to make sure it's not a false negative
```kotlin
typealias Dep = java.lang.Deprecated
@Dep
fun f() = Unit
```

Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:14]: add test for blocks
```kotlin
fun f(list: List<String>) {
    list.map @Suppress("x") { it.length }
}
```

Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:14]: add tests for expressions
```
val num = @Suppress("MagicNumber") 42

val x = @Suppress("UnnecessaryParentheses") (((1)+(2))) + 3
```

Comment (TWiStErRob) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:14]: ... I know, there are some weird places where you can place annotations 😅

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt (hunk 5 above)
Comment (ov7a) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:14]: Not sure if I solved this problem in a best way possible, but at least it works:) Thanks for great comments!

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+    @Test
+    fun `should report SuppressWarnings usages by default`() {
+        val code = """
+        @SuppressWarnings("unused")
```
Comment (cortinico) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:19]:
```suggestion
        @SuppressWarnings("unused")    
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+        import java.lang.annotation.Retention 
+        import java.lang.annotation.Documented 
+        import java.lang.annotation.Target
+        import java.lang.annotation.Repeatable
+        import java.lang.annotation.Inherited
+        import java.lang.annotation.RetentionPolicy
+        import java.lang.annotation.ElementType
+        import java.lang.Deprecated
+        @Deprecated
+        @Documented
```
Comment (cortinico) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:48]:
```suggestion
        @Documented
```

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+import org.jetbrains.kotlin.types.KotlinType
+
+/**
+ * This rule allows to set a list of forbidden annotations. This can be used to discourage the use
+ * of language annotations which do not require explicit import.
+ *
+ * <noncompliant>
+ * @@SuppressWarnings("unused")
+ * class SomeClass()
+ * </noncompliant>
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:32]: I would keep both a `compliant` and a `noncompliant` block here.
Maybe show the `@Deprecated` one or others

**Review by cortinico — APPROVED** [review]
LGTM 👍 Mostly minor nits

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt [hunk 11]
```diff
@@ -36,8 +36,8 @@ class ForbiddenAnnotationSpec(val env: KotlinCoreEnvironment) {
     @Test
     fun `should report annotations from java lang annotation package by default`() {
         val code = """
-        import java.lang.annotation.Retention 
-        import java.lang.annotation.Documented
```
Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt]: this will fail CI

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    override val issue = Issue(
+        javaClass.simpleName,
+        Severity.Style,
+        "Avoid using this annotation.",
+        Debt.FIVE_MINS
+    )
+
+    @Configuration(
+        "List of fully qualified annotation classes which are forbidden. " +
+            "For example, `kotlin.jvm.Transient`."
```
Comment (TWiStErRob) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:51]: This doc looks outdated since there's now `reason` too. Probably no need for `for example` since there are many examples in default config.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                "Use `kotlin.annotation.MustBeDocumented` instead.",
+            "java.lang.annotation.Target" to "it is a java annotation. Use `kotlin.annotation.Target` instead.",
+            "java.lang.annotation.Retention" to "it is a java annotation. Use `kotlin.annotation.Retention` instead.",
+            "java.lang.annotation.Repeatable" to "it is a java annotation. Use `kotlin.annotation.Repeatable` instead.",
+            "java.lang.annotation.Inherited" to "Kotlin does not support @Inherited annotation, " +
+                "see https://youtrack.jetbrains.com/issue/KT-22265",
+        )
+    ) { list ->
+        list.associate { it.value to Forbidden(it.value, it.reason) }
+    }
```
Comment (TWiStErRob) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:67]: Opinion: I think this can be simplified to `list.associateBy { it.value }`. It took me more to understand the extra private inner class than the win between `forbidden.value` and `forbidden.name`.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+            check(annotation, it)
+        }
+    }
+
+    override fun visitExpression(expression: KtExpression) {
+        super.visitExpression(expression)
+
+        if (annotations.isEmpty()) {
+            return
+        }
```
Comment (TWiStErRob) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:85]: Is this permature optimization? (same in above method) why would this rule be active if there are no configured annotations. @BraisGabin /@cortinico is there a way to easily check this during configuration phase? Can ConfigValidator be on a rule level? (btw. ConfigValidator has no [docs](https://detekt.dev/docs/next/introduction/extensions).)

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    private fun KtExpression.expressionTypeOrNull(): KotlinType? {
+        return if (bindingContext != BindingContext.EMPTY) {
+            bindingContext[BindingContext.EXPRESSION_TYPE_INFO, this]?.type
+        } else {
+            null
+        }
+    }
```
Comment (TWiStErRob) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:127]: `@RequiresTypeResolution` guarantees non-empty binding context. @BraisGabin this looks like another one for the rule-authors ruleset.

```suggestion
    private fun KtTypeReference.fqNameOrNull(): FqName? =
        bindingContext[BindingContext.TYPE, this]?.fqNameOrNull()

    private fun KtExpression.expressionTypeOrNull(): KotlinType? =
        bindingContext[BindingContext.EXPRESSION_TYPE_INFO, this]?.type
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+        val code = """
+        @Deprecated("unused", ReplaceWith("bar"))
+        fun foo() = "1234"
+        """.trimIndent()
+        val findings = ForbiddenAnnotation(
+            TestConfig(mapOf(ANNOTATIONS to listOf("kotlin.ReplaceWith")))
+        ).compileAndLintWithContext(env, code)
+        assertThat(findings).hasSize(1)
+            .hasStartSourceLocation(1, 23)
+            .hasTextLocations("ReplaceWith")
```
Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:209]: I think this is a redundant suppression, since there's only one occurrence of ReplaceWith

```suggestion
            .hasTextLocations("ReplaceWith")
```

Comment (TWiStErRob) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:209]: Same for most tests, hasTextLocations might be superior to other numeric validations (hasStartSourceLocation/hasSize) if the text is unique. Curious what others think.

**Review by TWiStErRob — APPROVED** [review]
Sorry, every time I look, I'm in a different mindset :)

Most comments are preference now, take them as you wish.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt (hunk 16 above)
Comment (ov7a) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt:209]: I thought so too, but it was really frustrating to get an exception that the text from the argument appears more than once in a snippet. I thought it kinda confusing to have half of the tests with text and half with numbers.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt (hunk 15 above)
Comment (TWiStErRob) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt]: leaving it open, curious what @BraisGabin thinks

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt (hunk 13 above)
Comment (ov7a) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:66]: I think for the sake of readability/clear types `Map<String, Forbidden>` is a bit better than `Map<String, String>`. The best variant would be something like `Map<String, Reason>` or even `Map<FqName, Reason>`, but this requires more work (unrelated to this PR). But if others think that `Forbidden` is redundant too, I'm ok with deleting it.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt (hunk 16 above)
Comment (TWiStErRob) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt]: > half of the tests with text and half with numbers.

IMO that just highlights even more the type of the test.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt (hunk 13 above)
Comment (TWiStErRob) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:66]: It's not going to be `<String, String>` you still get the same object, just with a different name:

```patch
Index: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt b/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt
--- a/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt	(revision 7ce3ebd0ba0cc58984ac3130847d10d23aa27254)
+++ b/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt	(date 1669413179980)
@@ -8,6 +8,7 @@
 import io.gitlab.arturbosch.detekt.api.Location
 import io.gitlab.arturbosch.detekt.api.Rule
 import io.gitlab.arturbosch.detekt.api.Severity
+import io.gitlab.arturbosch.detekt.api.ValueWithReason
 import io.gitlab.arturbosch.detekt.api.config
 import io.gitlab.arturbosch.detekt.api.internal.Configuration
 import io.gitlab.arturbosch.detekt.api.internal.RequiresTypeResolution
@@ -50,7 +51,7 @@
         "List of fully qualified annotation classes which are forbidden. " +
             "For example, `kotlin.jvm.Transient`."
     )
-    private val annotations: Map<String, Forbidden> by config(
+    private val annotations: Map<String, ValueWithReason> by config(
         valuesWithReason(
             "java.lang.SuppressWarnings" to "it is a java annotation. Use `Suppress` instead.",
             "java.lang.Deprecated" to "it is a java annotation. Use `kotlin.Deprecated` instead.",
@@ -63,7 +64,7 @@
                 "see https://youtrack.jetbrains.com/issue/KT-22265",
         )
     ) { list ->
-        list.associate { it.value to Forbidden(it.value, it.reason) }
+        list.associateBy { it.value }
     }
 
     override fun visitAnnotationEntry(annotation: KtAnnotationEntry) {
@@ -93,9 +94,9 @@
 
         if (forbidden != null) {
             val message = if (forbidden.reason != null) {
-                "The annotation `${forbidden.name}` has been forbidden: ${forbidden.reason}"
+                "The annotation `${forbidden.value}` has been forbidden: ${forbidden.reason}"
             } else {
-                "The annotation `${forbidden.name}` has been forbidden in the detekt config."
+                "The annotation `${forbidden.value}` has been forbidden in the detekt config."
             }
             val location = Location.from(element).let { location ->
                 location.copy(
@@ -108,8 +109,6 @@
         }
     }
 
-    private data class Forbidden(val name: String, val reason: String?)
-
     private fun KtTypeReference.fqNameOrNull(): FqName? {
         return if (bindingContext != BindingContext.EMPTY) {
             bindingContext[BindingContext.TYPE, this]?.fqNameOrNull()

```

Comment (TWiStErRob) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:66]: I sometimes do [`private typealias FullyQualifiedAnnotationName = String`](https://github.com/TWiStErRob/net.twisterrob.gradle/blob/5f3a9f7b71d54627dea85ec7f95b7e38d3d63c0c/quality/src/main/kotlin/net/twisterrob/gradle/quality/report/TableGenerator.kt#L6-L9) in my personal code, but that probably goes a bit too far in this project :)

Comment (ov7a) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:66]: Oh, missed that, thanks.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt (hunk 16 above)
Comment (TWiStErRob) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt]: Nice one! It actually answers one thing I wasn't sure of, in form of a test. The highlight includes the @ symbol :)

Note: because of the @ symbol, it should be possible to use text matchers everywhere, no?

Comment (ov7a) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt]: Yes, indeed. Only one place remains where there are actually two of them (annotations for expressions), but I guess it's ok to leave only one expression for test. 

Thank you, sorry for me being so careless:(

Comment (TWiStErRob) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotationSpec.kt]: No worries, I'm learning with you :)

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt (hunk 14 above)
Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt:85]: I agree, there's not need for this `if`. I don't care that much about it either... but if we have this `if`s we should also have a test to check what happen with `annotations` is empty. So I would say that it's better to remove it. But if you, @ov7a, prefer to keep it just add a test for it.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt (hunk 15 above)
Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt]: Yes, I think that any check of `bindingContext != BindingContext.EMPTY` or `bindingContext == BindingContext.EMPTY` should be flagged as a code smell. That should be an easy rule to implement.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt (hunk 14 above)
Comment (ov7a) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt]: I just removed it. I agree with @TWiStErRob that it would be nice to have some example of config validation.
